### ISSUE TITLE
[mlir][Vector][Affine] SuperVectortizer: Optimization for misaligned data.

### ISSUE THREAD

**Issue by codemzs** [llvm/llvm-project#86084]
Hello,

I’m reaching out to discuss a potential optimization in the SuperVectorizer, specifically regarding the handling of misaligned data during vectorization. In scenarios where the data size is not evenly divisible by the vector length, misalignment occurs, leading to the presence of garbage elements in the last iteration of a loop and this becomes important in cases where reduction occurs after the loop.

Currently, SuperVectorizer handles this by generating a mask in every iteration to avoid operations on these garbage values. While this ensures correctness, it introduces overhead by necessitating vector predication in all iterations (even though the mask is all 1s until the last iteration), which can significantly impact performance.

I propose considering an alternative approach, such as loop peeling or introducing a cleanup loop. This would involve processing the bulk of the data in the main loop without masking and then handling the remaining elements in a separate loop. Specifically, the cleanup loop would operate only on the last few elements, where masking is necessary to avoid garbage data processing.

This method could potentially reduce the overhead by limiting the use of masks to only the necessary part of the loop, avoiding the performance penalty associated with vector predication throughout the entire loop. The main benefits of this approach would be:

1. Reduced computational overhead by minimizing the use of costly predication logic.
2. Enhanced performance for loops dealing with misaligned data, particularly in reduction scenarios.
3. Cleaner and more efficient code generation, especially in tight loops where the overhead of masking in every iteration can be particularly pronounced.
 
I believe this optimization could lead to significant performance improvements in scenarios where misaligned vectorization is common. I would appreciate your thoughts on this suggestion and its feasibility within the current architecture of SuperVectorizer.

Thank you for considering this optimization proposal. If it makes sense I will be happy to implement it.

Thanks,
Zeeshan Siddiqui

CC: @dcaballe @sergei-grechanik

**Comment by sergei-grechanik** [comment]
Yes, it's an important optimization, but in my opinion it should be done in a separate pass. I know there is an scf loop peeling pass, but I don't known if it works for the IR produced by SuperVectorize. It's of course would be nice to have an affine loop peeling pass.
